common/src/main/java/io/netty/util/concurrent/ReentrantImmediateEventExecutor.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util.concurrent;
+
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+/**
+ * Executes {@link Runnable} objects in the caller's thread. If the {@link #execute(Runnable)} is reentrant it will be
+ * queued until the original {@link Runnable} finishes execution.
+ * <p>
+ * All {@link Throwable} objects thrown from {@link #execute(Runnable)} will be swallowed and logged. This is to ensure
+ * that all queued {@link Runnable} objects have the chance to be run.
+ */
+public final class ReentrantImmediateEventExecutor extends ImmediateEventExecutor {
```

Comment:
I would like to get input on this. I created a new class because `ImmediateEventExecutor` currently propagates exceptions thrown by the `Runnable` in the `execute` method. Now that we maintain a queue of `Runnables` it is not straightforward how to preserve this behavior (throw the first exception, last exception, etc..?). If we are OK with no longer propagating exceptions from `ImmediateEventExecutor`, or can agree on a sensible way to propagate when multiple Runnable may throw then we can combine this class and `ImmediateEventExecutor`